File: frontend/services/heatmap.service.ts
import { apiClient } from "./api.client";

export interface HeatmapPoint {
  lat: number;
  lng: number;
  count: number;
  categories: string[];
}

export interface HeatmapResponse {
  success: boolean;
  data: HeatmapPoint[];
  total: number;
  points: number;
}

export const heatmapService = {
  getHeatmapData: async (params: {
    category?: string;
    status?: string;
    municipality?: string;
    department?: string;
  }): Promise<HeatmapResponse> => {
    const queryParams = new URLSearchParams();
    
    if (params.category) queryParams.append('category', params.category);
    if (params.status) queryParams.append('status', params.status);
    if (params.municipality) queryParams.append('municipality', params.municipality);
    if (params.department) queryParams.append('department', params.department);
    
    const response = await apiClient.get<HeatmapResponse>(`/heatmap?${queryParams.toString()}`);
    return response;
  },
  
  getCategories: async (): Promise<{ success: boolean; categories: string[] }> => {
    const response = await apiClient.get<{ success: boolean; categories: string[] }>('/heatmap/categories');
    return response;
  }
};

File: mobile/lib/services/heatmap_service.dart
import 'api_client.dart';

class HeatmapService {
  final ApiClient _apiClient = ApiClient();

  Future<List<Map<String, dynamic>>> getHeatmapData({
    String? category,
    String? status,
    String? municipality,
    String? department,
  }) async {
    String endpoint = '/heatmap';
    final params = <String, String>{};
    if (category != null && category.isNotEmpty) {
      params['category'] = category;
    }
    if (status != null && status.isNotEmpty) params['status'] = status;
    if (municipality != null && municipality.isNotEmpty) {
      params['municipality'] = municipality;
    }
    if (department != null && department.isNotEmpty) {
      params['department'] = department;
    }
    if (params.isNotEmpty) {
      final query = params.entries.map((e) => '${e.key}=${e.value}').join('&');
      endpoint += '?$query';
    }
    final response = await _apiClient.get(endpoint);
    final data = response['data'] ?? response;
    final list = data is Map ? (data['points'] ?? data['data'] ?? []) : data;
    return (list as List).cast<Map<String, dynamic>>();
  }

  Future<List<String>> getCategories() async {
    final response = await _apiClient.get('/heatmap/categories');
    final cats = response['categories'] ?? response['data'] ?? [];
    return (cats as List).cast<String>();
  }
}
